### ISSUE TITLE
False positive in needless_pass_by_value on referenced fns

### ISSUE THREAD

**Issue by llogiq** [rust-lang-nursery/rust-clippy#1939]
The lint does not take into account that the interface may be restricted to a
value type when the function is used as rvalues or in arguments.

Example:

```
fn validate<F: FnMut(String) -> bool>(_: F) { unimplemented!(); }

fn take_string_by_value(x: String) -> bool { x.is_empty() }

fn main() {
    validate(take_string_by_value);
}
```

lints with

```
warning: this argument is passed by value, but not consumed in the function body
 --> src/main.rs:3:28
  |
3 | fn take_string_by_value(x: String) -> bool { x.is_empty() }
  |                            ^^^^^^ help: consider changing the type to: `&str`
  |
  = note: #[warn(needless_pass_by_value)] on by default
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_pass_by_value
```

However, adding a `&` to the `String` leads to a compilation error. The solution would be to check for references (we likely need this functionality elsewhere, so we may have it already)

**Comment by oli-obk** [comment]
ugh... We need to check all use sites and technically the moment the function is public we need to bail out, too, because external users' code will be broken by it.

I think this will make the lint much less useful, but I also don't see how to improve it.

Note: the lint is already a breaking change, because you need to pass `&foo` instead of `foo` to the function.

> Dear Rust,
> please consider automatically inserting an intermediate closure for argument derefs
> Yours
> Oli

**Comment by dtolnay** [comment]
Closing as a duplicate of #2434 (I know, that one is newer but it has more :+1: so it is the one that searchers are landing on).
